### PR TITLE
add basic auth support

### PR DESCRIPTION
**Pull request by reinsch82**
basically taken the code from [gradle wrapper](https://github.com/gradle/gradle/blob/master/subprojects/wrapper/src/main/java/org/gradle/wrapper/Download.java) 

This would be helpful in corporate environments with inhouse maven server with authentication enabled
(I happen to work in a company like that)
I've done this in my spare time which means no corporate copyrights are attached.

### CODE REVIEW

**Review by arbue — APPROVED** [review]

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java [hunk 1]
```diff
@@ -93,6 +87,54 @@ private void downloadInternal(URI address, File destination) throws Exception {
     }
   }
 
+  private void addBasicAuthentication(URI address, URLConnection connection) throws IOException {
+    String userInfo = calculateUserInfo(address);
+    if (userInfo == null) {
+        return;
+    }
+    if (!"https".equals(address.getScheme())) {
+       Logger.info("WARNING Using HTTP Basic Authentication over an insecure connection to download the Gradle distribution. Please consider using HTTPS.");
```
Comment (mosabua) [inline src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: Maven distribution

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java [hunk 2]
```diff
@@ -93,6 +87,54 @@ private void downloadInternal(URI address, File destination) throws Exception {
...
+              Method encodeMethod = loader.loadClass("javax.xml.bind.DatatypeConverter").getMethod("printBase64Binary", byte[].class);
+              return (String) encodeMethod.invoke(null, new Object[]{userInfo.getBytes("UTF-8")});
+          } catch (Exception java5OrEarlier) {
+              throw new RuntimeException("Downloading Gradle distributions with HTTP Basic Authentication is not supported on your JVM.", java5OrEarlier);
+          }
+      }
+  }
+
+  private String calculateUserInfo(URI uri) {
+    String username = System.getProperty("maven.wrapperUser");
```
Comment (mosabua) [inline src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: Can you change this to use environment variables MVNW_USER and MVNW_PASSWORD to be consistent with the already used MVNW_VERBOSE and add documentation in the readme file into this PR?

I am hoping to release a new version of the wrapper soon..

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java [hunk 3]
```diff
@@ -16,16 +16,9 @@
...
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.net.Authenticator;
-import java.net.PasswordAuthentication;
-import java.net.URI;
-import java.net.URL;
-import java.net.URLConnection;
+import java.io.*;
```
Comment (mosabua) [inline src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: please dont collapse the imports

**Comment by mosabua** [comment]
I think this is not needed if https://github.com/takari/maven-wrapper/pull/86/files is fixed and merged. Right? As it stands this is kinda incomplete on its own.

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java (hunk 1 above)
Comment (reinsch82) [reply src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: fixed

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java (hunk 2 above)
Comment (reinsch82) [reply src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: added

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java (hunk 3 above)
Comment (reinsch82) [reply src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: fixed

**Comment by reinsch82** [comment]
To my understanding these 2 PRs complete each other.
This one gets the Maven distribution with Basic auth.
The other one gets the maven wrapper jar with Basic auth.
We could use both I think.
Do I miss something in #86 ?
This on it's own does make sense. I currently can't see how just getting the wrapper would make sense on it's own. Where would I get the distribution from?

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java [hunk 4]
```diff
@@ -93,6 +99,54 @@ private void downloadInternal(URI address, File destination) throws Exception {
...
+          Method getEncoderMethod = loader.loadClass("java.util.Base64").getMethod("getEncoder");
+          Method encodeMethod = loader.loadClass("java.util.Base64$Encoder").getMethod("encodeToString", byte[].class);
+          Object encoder = getEncoderMethod.invoke(null);
+          return (String) encodeMethod.invoke(encoder, new Object[]{userInfo.getBytes("UTF-8")});
+      } catch (Exception java7OrEarlier) {
+          try {
+              Method encodeMethod = loader.loadClass("javax.xml.bind.DatatypeConverter").getMethod("printBase64Binary", byte[].class);
+              return (String) encodeMethod.invoke(null, new Object[]{userInfo.getBytes("UTF-8")});
+          } catch (Exception java5OrEarlier) {
+              throw new RuntimeException("Downloading Maven   distributions with HTTP Basic Authentication is not supported on your JVM.", java5OrEarlier);
```
Comment (sepe81) [inline src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: Two duplicate spaces after `Downloading Maven` could be removed

File: src/main/java/org/apache/maven/wrapper/DefaultDownloader.java [hunk 5]
```diff
@@ -93,6 +99,54 @@ private void downloadInternal(URI address, File destination) throws Exception {
     }
   }
 
+  private void addBasicAuthentication(URI address, URLConnection connection) throws IOException {
+    String userInfo = calculateUserInfo(address);
+    if (userInfo == null) {
+        return;
+    }
+    if (!"https".equals(address.getScheme())) {
+       Logger.info("WARNING Using HTTP Basic Authentication over an insecure connection to download the Maven distribution. Please consider using HTTPS.");
```
Comment (sepe81) [inline src/main/java/org/apache/maven/wrapper/DefaultDownloader.java]: It looks a bit confusing to put a WARNING into info logging. Why not use warn logging?

**Review by sepe81 — CHANGES_REQUESTED** [review]

**Review by sepe81 — APPROVED** [review]

File: .mvn/wrapper/maven-wrapper.properties [hunk 6]
```diff
@@ -1,3 +1,3 @@
-distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.0/apache-maven-3.6.0-bin.zip
```
Comment (sepe81) [inline .mvn/wrapper/maven-wrapper.properties:1]: Possibly this PR branch needs to be rebased with master b/c some changes are already present there.

see https://github.com/takari/maven-wrapper/commit/b93c46869a626d95af1908713580d88fc349937c

**Pull request merged**